src/shell/shell.cpp

```diff
@@ -70,67 +70,129 @@ DOS_Shell::DOS_Shell()
 	}
 }
 
-void DOS_Shell::GetRedirection(char *line,
-                               std::string &in_file,
-                               std::string &out_file,
-                               std::string &pipe_file,
-                               bool *append)
+// This function gets redirection targets from the given shell command line and
+// returns the results as a struct. The results include the redirection targets
+// as well as the processed command line with the targets stripped off.
+//
+// Note that real MS-DOS is quite nuanced in its whitespace handling:
+// - 'echo 1>out.txt' produces a 3-byte file: '1CRLF'
+// - 'echo 1 > out.txt' produces a 4-byte file: '1 CRLF'
+// - 'echo 1 >out.txt ' produces a 5-byte file: '1  CRLF'
+// This behavior is replicated here; see the shell_redirection_tests for more
+// examples.
+//
+std::optional<DOS_Shell::RedirectionResults> DOS_Shell::GetRedirection(
+        const std::string_view line)
 {
-	char *line_read = line;
-	char *line_write = line;
-	char character = 0;
-	bool quote = false;
-	size_t found = 0;
-	size_t temp_len = 0;
-	std::string redir = "";
-	std::string find_chars = "";
-	std::string *output;
-	*append = false;
-	while ((character = *line_read++)) {
-		if (quote && character != '"') { /* don't parse redirection
-			                            within quotes. Not perfect
-			                            yet. Escaped quotes will
-			                            mess the count up */
-			*line_write++ = character;
-			continue;
+	// Returned results and quick-access references to members.
+	RedirectionResults results = {};
+	auto& [processed_line, in_file, out_file, pipe_target, is_appending] = results;
+
+	// The regex splits the line into quoted and redirection groups
+	static const std::regex re(R"(("[^"]*"\s*)|([<>|]|>>|<<)\s*([^<>| ]+)(\s*))",
+	                           std::regex::optimize);
+	//
+	//    ("[^"]*"\s*)  # Group 1: double-quoted text
+	//    |             # -or-
+	//    (             # Group 2: the redirection tokens, either
+	//        [<>|]     # Single character tokens
+	//        |         # -or-
+	//        >>|<<     # Double character tokens
+	//    )
+	//    \s*           # Whitespace after the redirection token
+	//    ([^<>| ]+)    # Group 3: target file or program name
+	//    (\s*)         # Group 4: the target's tail whitespace
```

Comment:
Give this man an award, please 💯 Yeah, regexes need to be documented like that, otherwise you're in write-only mode quickly 😏

Reply:
> otherwise you're in write-only mode quickly

That's for sure!

The comment is still not as 'DRY' as I'd like.. I tried hard to document the actual statement on the fly.. like:

```c++
regex(
  R("...") // Group 1 comment
  R("...") // Group 2 comment
  ... 
);
```

But that approach changed the behaviour, so I had to go with the single line.